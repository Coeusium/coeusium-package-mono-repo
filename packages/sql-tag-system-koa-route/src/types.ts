export type { ConnectionInfo, Tag } from 'sql-tag-system';
