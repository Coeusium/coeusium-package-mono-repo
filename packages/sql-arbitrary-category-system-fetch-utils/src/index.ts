export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
