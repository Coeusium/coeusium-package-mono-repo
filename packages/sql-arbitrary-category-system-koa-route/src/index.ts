export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
