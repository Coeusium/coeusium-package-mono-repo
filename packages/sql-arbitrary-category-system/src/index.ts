export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
