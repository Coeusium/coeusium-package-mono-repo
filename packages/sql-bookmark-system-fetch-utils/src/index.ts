export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
