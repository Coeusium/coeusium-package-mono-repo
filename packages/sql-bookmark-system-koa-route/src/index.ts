export * from './sql-bookmark-system-koa-route';
export type { ConnectionInfo, Bookmark } from './types';
