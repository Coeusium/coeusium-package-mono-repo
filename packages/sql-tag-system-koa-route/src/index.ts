export * from './sql-tag-system-koa-route';
export type { ConnectionInfo, Tag } from './types';
