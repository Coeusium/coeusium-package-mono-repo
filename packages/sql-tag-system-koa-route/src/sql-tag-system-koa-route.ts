import Router from '@koa/router';
import {
  addTag,
  deleteTag,
  getTag,
  listTags,
  tagIdExists,
} from 'sql-tag-system';
import compose from 'koa-compose';
import { validateQueryParamsAll } from '@almaclaine/koa-utils';
import { ConnectionInfo } from './types';

export { destroy, setupTagSystem } from 'sql-tag-system';

export function createGetTagHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getTag(dbInfo, id);
  };
}

export function createGetTagListHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { limit, page } = ctx.request.query;
    ctx.body = await listTags(dbInfo, page, limit);
  };
}

export function createTagGetHandler(dbInfo: ConnectionInfo) {
  const getTagHandler = createGetTagHandler(dbInfo);
  const listTagHandler = createGetTagListHandler(dbInfo);
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    if (id) {
      await getTagHandler(ctx, next);
    } else {
      await listTagHandler(ctx, next);
    }
  };
}

export function createPostTagHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { tag } = ctx.request.query;
    const id = await addTag(dbInfo, tag);
    ctx.body = { id };
  };
}

export function createTagDeleteHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    const exists = await tagIdExists(dbInfo, id);
    if (!exists) {
      ctx.status = 412;
      ctx.body = { error: 'No tag exists with id ' + id };
      return;
    }

    await deleteTag(dbInfo, id);
    ctx.body = {};
  };
}

export function getRouter(dbInfo: ConnectionInfo) {
  const router = new Router();

  const tagGetHandler = createTagGetHandler(dbInfo);
  router.get('/', tagGetHandler);

  const tagDeleteHandler = createTagDeleteHandler(dbInfo);
  router.delete(
    '/',
    compose([validateQueryParamsAll(['id']), tagDeleteHandler]),
  );

  const postTagHandler = createPostTagHandler(dbInfo);
  router.post('/', compose([validateQueryParamsAll(['tag']), postTagHandler]));

  return router;
}

export function getTagScopedRouter(dbInfo: ConnectionInfo) {
  const router = new Router();
  const routes = getRouter(dbInfo);
  router.use('/tag', routes.routes(), routes.allowedMethods());
  return router;
}
