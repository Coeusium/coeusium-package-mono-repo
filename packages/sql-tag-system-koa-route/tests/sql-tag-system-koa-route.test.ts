import * as sqlTagSystem from '../src';

describe('sql-tag-system-koa-route-test', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlTagSystem);
    expect(keys.length).toBe(9);
    expect(keys).toMatchSnapshot();
  });
});
