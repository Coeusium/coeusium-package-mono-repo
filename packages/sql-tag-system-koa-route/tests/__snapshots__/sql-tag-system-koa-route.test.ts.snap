// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-tag-system-koa-route-test Has Correct API 1`] = `
Array [
  "destroy",
  "setupTagSystem",
  "createGetTagHandler",
  "createGetTagListHandler",
  "createTagGetHandler",
  "createPostTagHandler",
  "createTagDeleteHandler",
  "getRouter",
  "getTagScopedRouter",
]
`;
